src/grt/src/fastroute/src/maze.cpp

```diff
@@ -61,6 +62,250 @@ static int right_index(int i)
   return 2 * i + 2;
 }
 
+bool FastRouteCore::checkTree(const int net_id)
+{
+  TreeEdge* treeedges = sttrees_[net_id].edges;
+  const int deg = sttrees_[net_id].deg;
+
+  // group all edges that crosses the same position
+  std::unordered_map<std::pair<short, short>,
+                     std::vector<int>,
+                     boost::hash<std::pair<int, int>>>
+      position_to_edges_map;
+
+  auto cmp = [&](int a, int b) { return treeedges[a].len > treeedges[b].len; };
+  std::map<int, std::vector<std::pair<short, short>>, decltype(cmp)>
+      edges_to_blocked_pos_map(cmp);
+  for (int edgeID = 0; edgeID < 2 * deg - 3; edgeID++) {
+    TreeEdge* treeedge = &(treeedges[edgeID]);
+    if (treeedge->len > 0) {
+      int routeLen = treeedge->route.routelen;
+      const std::vector<short>& gridsX = treeedge->route.gridsX;
+      const std::vector<short>& gridsY = treeedge->route.gridsY;
+      const std::pair<short, short> pos_0 = {gridsX[0], gridsY[0]};
+      position_to_edges_map[pos_0].push_back(edgeID);
+
+      for (int i = 1; i <= routeLen; i++) {
+        const std::pair<short, short> pos_i = {gridsX[i], gridsY[i]};
+        position_to_edges_map[pos_i].push_back(edgeID);
+      }
+    }
+  }
+
+  // for each position, check if there is an edge that don't share
+  // a node with the remaining edges
+  for (auto const& [position, edges] : position_to_edges_map) {
+    for (int edgeID : edges) {
+      if (checkOverlapEdge(net_id, edgeID, edges)) {
+        edges_to_blocked_pos_map[edgeID].push_back(
+            {position.first, position.second});
+      }
+    }
+  }
+
+  // fix the longest edge and break the loop
+  for (auto& [edge, blocked_positions] : edges_to_blocked_pos_map) {
+    fixOverlappingEdge(net_id, edge, blocked_positions);
+    break;
+  }
+
+  return true;
+}
+
+bool FastRouteCore::checkOverlapEdge(const int net_id,
+                                     const int edge_id,
+                                     const std::vector<int>& edges)
+{
+  TreeEdge* treeedges = sttrees_[net_id].edges;
+  TreeEdge* treeedge = &(treeedges[edge_id]);
+
+  int n1 = treeedge->n1;
+  int n2 = treeedge->n2;
+  int n1a = treeedge->n1a;
+  int n2a = treeedge->n2a;
+
+  int common_positions = 0;
+  for (int ed : edges) {
+    if (ed != edge_id) {
+      int ed_n1 = treeedges[ed].n1;
+      int ed_n2 = treeedges[ed].n2;
+      int ed_n1a = treeedges[ed].n1a;
+      int ed_n2a = treeedges[ed].n2a;
+      if ((ed_n1a == n1a) || ((ed_n1a == n2a)) || (ed_n1 == n1)
+          || (ed_n1 == n2)) {
+        common_positions++;
```

Comment:
there is no point to continue the loop as you only care zero/non-zero not the final count.